chain/network/src/peer_manager/peer_manager_actor.rs

```diff
@@ -1063,90 +1120,84 @@ impl PeerManagerActor {
         }
+        // Add whitelisted nodes to the safe set.
+        safe_set.extend(whitelisted_peers.into_iter().cloned());
 
-        if self.config.archive
-            && (self.connected_peers.values())
-                .filter(|connected_peer| connected_peer.full_peer_info.chain_info.archival)
-                .count()
-                <= self.config.archival_peer_connections_lower_bound as usize
+        // If there is not enough outbound peers, add them to the safe set.
+        let outbound_peers = filter_peers(&|p| p.peer_type == PeerType::Outbound);
+        if outbound_peers.len() + self.outgoing_peers.len()
+            <= self.config.minimum_outbound_peers as usize
         {
-            for (peer, active) in self.connected_peers.iter() {
-                if active.full_peer_info.chain_info.archival {
-                    safe_set.insert(peer);
-                }
-            }
+            safe_set.extend(outbound_peers.into_iter().cloned());
         }
 
-        // Find all recent connections
-        let mut recent_connections = (self.connected_peers.iter())
-            .filter_map(|(peer_id, active)| {
-                if active.last_time_received_message.elapsed() < self.config.peer_recent_time_window
-                {
-                    Some((peer_id.clone(), active.connection_established_time))
-                } else {
-                    None
-                }
-            })
-            .collect::<Vec<_>>();
-
-        // Sort by established time
-        recent_connections.sort_by(|(_, established_time_a), (_, established_time_b)| {
-            established_time_a.cmp(established_time_b)
-        });
+        // If there is not enough archival peers, add them to the safe set.
+        if self.config.archive {
+            let archival_peers = filter_peers(&|p| p.full_peer_info.chain_info.archival);
+            if archival_peers.len() <= self.config.archival_peer_connections_lower_bound as usize {
+                safe_set.extend(archival_peers.into_iter().cloned());
+            }
+        }
 
-        // Take remaining peers
-        for (peer_id, _) in recent_connections
+        // Find all recently active peers.
+        let mut active_peers: Vec<_> = self
+            .connected_peers
             .iter()
-            .take((self.config.safe_set_size as usize).saturating_sub(safe_set.len()))
-        {
-            safe_set.insert(peer_id);
+            .filter(|(_, p)| {
+                p.last_time_received_message.elapsed() < self.config.peer_recent_time_window
+            })
+            .collect();
+        // Sort by established time.
+        active_peers.sort_by_key(|(_, p)| p.connection_established_time);
+        // Saturate safe set with recently active peers.
+        let set_limit = self.config.safe_set_size as usize;
+        for (id, _) in active_peers {
+            if safe_set.len() >= set_limit {
+                break;
+            }
+            safe_set.insert(id.clone());
         }
```

Comment:
Probably wrap this whole thing in `if safe_set.len() < set_limit { ... }` so that we don’t waste time going through peers and sorting them if we won’t add any of the peers to safe set in the end.